### PR TITLE
ytrafo pipeops

### PR DESCRIPTION
**Pull request by sumny**
See #322 and #364 
This is still WIP.
Adds abstract `PipeOpInvertiblePreproc` from which you can inherit your own trafo `PipeOp`. Toy example is given in `PipeOpSimpleTrafo`. Together with `PipeOpInverter` this allows to build a `Graph` with a `Learner` in the middle, so that the target is transformed during training and retransformed for prediction. Docu is drafted; tests missing. Let's discuss this first

### CODE REVIEW

File: R/PipeOpTrafo.R [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    initialize = function(id = "logtrafo") {
+      super$initialize(id = id)
+    },
+
+    train_target = function(task) {
+      self$state = list()
+      new_target = log(task$data(cols = task$target_names))
+      names(new_target) = paste0("log.", task$target_names)
+      task$cbind(new_target)
+      private$update_target(task, new_target = names(new_target), new_type = NULL)
```
Comment (mb706) [inline R/PipeOpTrafo.R]: how about a signature
```r
private$update_target(task, new_target, new_target_name = NULL, new_type = NULL)
```
and calling it here as
```r
new_target = log(task$data(cols = task$target_names))
private$update_target(task, new_target = new_target,
  new_target_name = paste0("log.", task$target_names))
```

**Comment by pfistfl** [comment]
Please test that you can construct an arbitrary `log(base = X)` -> invert = `x ^ base` trafo.

File: R/PipeOpTrafo.R [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+    inverter = function(prediction, predict_phase_control) {
+      # function that inverts the predictions and returns a Prediction object
+      stop("Abstract.")
+    }
+  ),
+  private = list(
+    update_target = function(task, new_target, new_type = NULL, ...) {
+      type = if (is.null(new_type)) task$task_type else assert_subset(new_type, mlr_reflections$task_types$type, empty.ok = FALSE)
+      get(mlr_reflections$task_types[type]$task)$new(id = task$id, backend = task$backend, target = new_target, ...)
```
Comment (pfistfl) [inline R/PipeOpTrafo.R]: See `convert_target` from hurdle models branch;  do not convert task if task_type stays the same.

File: R/PipeOpTrafo.R [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    invert_help = function(predict_phase_control) {
+      function(inputs) {
+        assert_list(inputs, len = 1L, types = "Prediction")
+        self$inverter(inputs[[1L]], predict_phase_control)
+      }
+    },
+
+    set_target_to_NA_and_possibly_convert_task_type = function(task) {
+      #FIXME: convert the task target to NA and convert the task type?
```
Comment (pfistfl) [inline R/PipeOpTrafo.R]: I guess this should just set the  `col_roles` of the original target to 0.

File: R/PipeOpTrafo.R [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+    predict_internal = function(inputs) {
+      list(inputs[[1L]](inputs[-1L]))
+    }
+  )
+)
+
+mlr_pipeops$add("inverter", PipeOpInverter)
+
+#' @title PipeOpLogtrafo
```
Comment (pfistfl) [inline R/PipeOpTrafo.R]: PipeOpSimpleTrafo:
- following params in its `param_set`: `trafo`, `inverter`, `new_target_name`, `new_task_type`
- `trafo` and `inverter` default to `identity`
- `new_target_name` defaults to `paste0("trafo_", xxold_target_name)`
- `new_task_type` defaults to `NULL` (keep old task type)

Comment (pfistfl) [reply R/PipeOpTrafo.R]: then `PipeOpLogTrafo` could inherit from this and just define different `trafo` and `inverter`.

Comment (sumny) [reply R/PipeOpTrafo.R:162]: done, maybe we discuss whether another inheritance from `PipeOpSimpleTrafo` is needed for `PipeOpLogTrafo` (it is pretty easy to set this up already in `PipeOpSimpleTrafo`)

**Comment by sumny** [comment]
> Please test that you can construct an arbitrary `log(base = X)` -> invert = `x ^ base` trafo.

Added an example in `PipeOpSimpleTrafo`.

**Comment by pfistfl** [comment]
Could you also add the `ppl` shortcut and also document this?

File: R/PipeOpNewTarget.R [hunk 5]
```diff
@@ -0,0 +1,43 @@
+#' Convert a task from its type to another.
+#'
+#' The task's target is replaced by a different column from the data.
+#' #FIXME: This needs detailed tests.
```
Comment (pfistfl) [inline R/PipeOpNewTarget.R:4]: + merge relevant tests from the old PR.

File: R/PipeOpNewTarget.R [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+#'   A [`Task`][mlr3::Task] to be converted.
+#' @param new_target `character(1)`\cr
+#'   New target to be set, must be a column in the `intask` data.
+#'   If `NULL`, no new target is set, and task is converted as-is.
+#' @param new_type `character(1)`\cr
+#'   The new task type. Must be in `mlr_reflections$task_types`.
+#'   If `NULL` (default), a new task with the same task_type is created.
+#' @param drop_original_target `logical(1)`\cr
+#'   If `FALSE` (default), the original target is added as a feature.
+#'   Otherwise the original target is assigned no col_role, i.e. dropped.
```
Comment (pfistfl) [inline R/PipeOpNewTarget.R:15]: We assign `character(0)` in order to drop col_roles

File: R/PipeOpTrafo.R [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+#'po$predict(list(task))
+#'
+#'g = Graph$new()
+#'g$add_pipeop(po)
+#'g$add_pipeop(LearnerRegrRpart$new())
+#'g$add_pipeop(PipeOpInverter$new())
+#'g$add_edge(src_id = "logtrafo", dst_id = "inverter", src_channel = 1, dst_channel = 1)
+#'g$add_edge(src_id = "logtrafo", dst_id = "regr.rpart", src_channel = 2, dst_channel = 1)
+#'g$add_edge(src_id = "regr.rpart", dst_id = "inverter", src_channel = 1, dst_channel = 2)
+#'
```
Comment (pfistfl) [inline R/PipeOpTrafo.R:305]: Also show the `ppl("targettrafo", ...)` version here.

**Comment by sumny** [comment]
Renamed the `PipeOp`s, added tests, and also a `ppl` that allows for an easier build around a `PipeOpLearner` or `Graph` (where the last `PipeOp` is a `PipeOpLearner`). Also I am more strict regarding the `Task` and `Prediction` within `PipeOpTargetTrafoSimple` - I think this is currently only useful for regression tasks where the target is a (continous) numerical variable (but you can prove me wrong). This is now ready for a final review!

File: R/PipeOpTrafo.R [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+    initialize = function(id, param_set = ParamSet$new(), param_vals = list(), packages = character(0), tags = NULL) {
+      super$initialize(id = id, param_set = param_set, param_vals = param_vals,
+        input = data.table(name = "input", train = "Task", predict = "Task"),
+        output = data.table(name = c("fun", "output"), train = c("NULL", "Task"), predict = c("function", "Task")),
+        packages = packages, tags = tags
+      )
+    },
+
+    train_target = function(task) {
+      # return a modified task and set the state
```
Comment (pfistfl) [inline R/PipeOpTrafo.R:109]: do we still need all the comments?

File: R/PipeOpTrafo.R [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+#'
+#' @section Internals:
+#' Overloads [`PipeOpTargetTrafo`]'s `train_target()`, `train_invert()` and `inverter()`.
+#' Should be used in combination with [`PipeOpTargetInverter`].
+#'
+#' @section Methods:
+#' Only methods inherited from [`PipeOpTargetTrafo`]/[`PipeOp`].
+#'
+#' @examples
+#'library(mlr3)
```
Comment (pfistfl) [inline R/PipeOpTrafo.R:291]: indent here

File: R/mlr_graphs_elements.R [hunk 10]
```diff
@@ -243,3 +244,67 @@ pipeline_branch = function(graphs, prefix_branchops = "", prefix_paths = FALSE)
 }
 
 mlr_graphs$add("branch", pipeline_branch)
+
+
+
+#' @title Transform A Numerical Target And Later Invert The Transformation
```
Comment (pfistfl) [inline R/mlr_graphs_elements.R:250]: Transform and re-transform the target variable

File: R/mlr_graphs_elements.R [hunk 11]
```diff
@@ -243,3 +244,67 @@ pipeline_branch = function(graphs, prefix_branchops = "", prefix_paths = FALSE)
...
+#' @name mlr_graphs_targettrafo
+#' @description
+#' Creates a [`Graph`] that transforms a numerical target of a [`TaskRegr`][mlr3::TaskRegr] during
```
Comment (pfistfl) [inline R/mlr_graphs_elements.R:253]: I would drop the numerical, this works for any target.

Comment (pfistfl) [inline R/mlr_graphs_elements.R:253]: Wraps a graph

File: R/mlr_graphs_elements.R [hunk 12]
```diff
@@ -243,3 +244,67 @@ pipeline_branch = function(graphs, prefix_branchops = "", prefix_paths = FALSE)
...
+#')
+#'g$add_pipeop(LearnerRegrRpart$new())
+#'g$add_pipeop(PipeOpTargetInverter$new())
+#'g$add_edge(src_id = "targettrafosimple", dst_id = "targetinverter",
+#'  src_channel = 1, dst_channel = 1)
+#'g$add_edge(src_id = "targettrafosimple", dst_id = "regr.rpart",
+#'  src_channel = 2, dst_channel = 1)
+#'g$add_edge(src_id = "regr.rpart", dst_id = "targetinverter",
+#'  src_channel = 1, dst_channel = 2)
+pipeline_targettrafo = function(graph) {
```
Comment (pfistfl) [inline R/mlr_graphs_elements.R:291]: signature should be `function(graph, trafo_pipeop = NULL, id_prefix = "")`
If  `trafo_pipeop = NULL` uses `PipeOpTargetTrafoSimple`, otherwise a custom target trafo can be provided.
Additionally prefixes the `id` of trafo and inverter with `id_prefix`. 
Otherwise, this could lead to issues as duplicate id's are not allowed.

Test that this works for custom id's aswell.

File: R/PipeOpNewTarget.R [hunk 13]
```diff
@@ -0,0 +1,43 @@
+#' Convert a task from its type to another.
```
Comment (pfistfl) [inline R/PipeOpNewTarget.R:1]: File Name is wrong

File: tests/testthat/test_pipeop_newtarget.R [hunk 14]
```diff
@@ -0,0 +1,20 @@
+test_that("convert_task reconstructs task", {
```
Comment (pfistfl) [inline tests/testthat/test_pipeop_newtarget.R:1]: could you copy all relevant tests?  (Or possibly test a few more edge cases).

In theory, this function should also be able to just change iris into a regr task by predicting `Sepal.Length` instead of `Species` etc. guess we  can do a few more sensible tests.

File: tests/testthat/test_pipeop_targettrafosimple.R [hunk 15]
```diff
@@ -0,0 +1,74 @@
+context("PipeOpTargetTrafoSimple")
```
Comment (pfistfl) [inline tests/testthat/test_pipeop_targettrafosimple.R:1]: So the following currently breaks, as it would require a `PredictionClassif` -> `PredictionRegr` conversion
- Task: `boston_housing`, `trafo = x > 20`,  inverter if (x) 19 else 21 (then we'd need a classif learner)

While the following might work. This is actually somewhat relevant.
- Task: `boston_housing`,`trafo = x > 20`,  inverter `identity`? This should probably be possible.

Comment (pfistfl) [reply tests/testthat/test_pipeop_targettrafosimple.R:1]: Could we test that and possibly add a fixme that we might somehow enable the first option in the future?

File: tests/testthat/test_pipeop_targettrafosimple.R [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+ dat$medv = log(dat$medv, base = 2)
+ task_log = TaskRegr$new("boston_housing_log", backend = dat, target = "medv")
+
+ learner = LearnerRegrRpart$new()
+ learner$train(task_log)
+
+ learner_predict_out = learner$predict(task_log)
+ expect_equal(2 ^ learner_predict_out$truth, predict_out[[1L]]$truth)
+ expect_equal(2 ^ learner_predict_out$response, predict_out[[1L]]$response)
+})
```
Comment (pfistfl) [inline tests/testthat/test_pipeop_targettrafosimple.R:74]: We currently have no checks for the state, i.e. whether learning and applying a state possibly works.

I guess we could do one `PipeOp` that covers this in the PR's comments.

**Comment by pfistfl** [comment]
We currently can not test that the `state` actually works as we have no `PipeOp` that uses it.

I will describe a `PipeOp` we should implement here in order to showcase how such pipeops are implemented and to test that this works in general.
```
PipeOpTargetTrafoScaleRange
...
basically does the same as `PipeOpScaleRange`, i.e. stores range trafo to state.
additionally, it knows how to retransform scaled data.
```

File: tests/testthat/test_pipeop_targettrafosimple.R (hunk 16 above)
Comment (sumny) [reply tests/testthat/test_pipeop_targettrafosimple.R:74]: yes, seems like a good idea!

File: tests/testthat/test_pipeop_targettrafosimple.R (hunk 15 above)
Comment (sumny) [reply tests/testthat/test_pipeop_targettrafosimple.R:1]: I added two more tests here, one that goes from  `PredictionClassif` -> `PredictionRegr` and one the other way. Both are not the most meaningful ones but it is possible. For most cases to actually be meaningful the user will rely on setting the `$state` and rely on this for the inversion (like statet by you below).

**Comment by sumny** [comment]
> We currently can not test that the `state` actually works as we have no `PipeOp` that uses it.
> 
> I will describe a `PipeOp` we should implement here in order to showcase how such pipeops are implemented and to test that this works in general.
> 
> ```
> PipeOpTargetTrafoScaleRange
> ...
> basically does the same as `PipeOpScaleRange`, i.e. stores range trafo to state.
> additionally, it knows how to retransform scaled data.
> ```

I added a `PipeOpTargetTrafoScaleRange` and tests. Seems to work nice. Had to add a `set_state()` method to `PipeOpTargetTrafo` that can be overloaded that makes such trafos way easier. `set_state()` is called a single time in `.train()`, before `train_task()` is evaluated so you can then rely on the `$state` both during training and prediction.
